src/States.cpp

```diff
@@ -0,0 +1,291 @@
+                               { return vec.x==0.0 && vec.y==0.0 && vec.z==0.0; });
+    return !allZero;
+  }
+
+
+  ale::State States::getState(double time, ale::interpolation interp) const {
+    std::vector<double> xs, ys, zs, vxs, vys, vzs;
+    int lowerBound = ale::interpolationIndex(m_ephemTimes, time);
+    std::vector<double> interpTimes = {};
+    int interpStart;
+    int interpStop;
+
+    if (m_ephemTimes.size() <= 3) {
+      interpStart = 0;
+      interpStop = m_ephemTimes.size() - 1;
+    }
+    else if (lowerBound == 0) {
+      interpStart = lowerBound;
+      interpStop = lowerBound + 3;
+    }
+    else if (lowerBound == m_ephemTimes.size() - 1) {
+      interpStart = lowerBound - 3;
+      interpStop = lowerBound;
+    }
+    else if (lowerBound == m_ephemTimes.size() - 2) {
+      interpStart = lowerBound - 2;
+      interpStop = lowerBound + 1;
+    }
+    else {
+      interpStart = lowerBound - 1;
+      interpStop = lowerBound + 2;
+    }
+
+    ale::State state;
+
+    for (int i = interpStart; i <= interpStop; i++) {
+      state = m_states[i];
+      interpTimes.push_back(m_ephemTimes[i]);
+      xs.push_back(state.position.x);
+      ys.push_back(state.position.y);
+      zs.push_back(state.position.z);
+      vxs.push_back(state.velocity.x);
+      vys.push_back(state.velocity.y);
+      vzs.push_back(state.velocity.z);
+    }
+
+    ale::Vec3d position, velocity;
+
+    if ( interp == LINEAR || (interp == SPLINE && !hasVelocity())) {
+      position = {interpolateState(xs,  interpTimes, time, interp, 0),
+                  interpolateState(ys,  interpTimes, time, interp, 0),
+                  interpolateState(zs,  interpTimes, time, interp, 0)};
+
+      velocity = {interpolateState(vxs, interpTimes, time, interp, 0),
+                  interpolateState(vys, interpTimes, time, interp, 0),
+                  interpolateState(vzs, interpTimes, time, interp, 0)};
+    }
+    else if (interp == SPLINE && hasVelocity()){
+      // Do hermite spline if velocities are available
+      double baseTime = (m_ephemTimes.at(0) + m_ephemTimes.at(m_ephemTimes.size() - 1)) / 2.;
+      double timeScale = 1.0;
+
+      std::vector<double> scaledEphemTimes;
+      for(unsigned int i = 0; i < interpTimes.size(); i++) {
+        scaledEphemTimes.push_back((interpTimes[i] - baseTime) / timeScale);
+      }
+      double sTime = (time - baseTime) / timeScale;
+      position.x = ale::evaluateCubicHermite(sTime, vxs, scaledEphemTimes, xs);
+      position.y = ale::evaluateCubicHermite(sTime, vys, scaledEphemTimes, ys);
+      position.z = ale::evaluateCubicHermite(sTime, vzs, scaledEphemTimes, zs);
+
+      velocity.x = ale::evaluateCubicHermiteFirstDeriv(sTime, vxs, scaledEphemTimes, xs);
+      velocity.y = ale::evaluateCubicHermiteFirstDeriv(sTime, vys, scaledEphemTimes, ys);
+      velocity.z = ale::evaluateCubicHermiteFirstDeriv(sTime, vzs, scaledEphemTimes, zs);
```

Comment:
These are technically incorrect, but because our time scaling has been 1, it hasn't shown up yet. We're taking a derivative in scaled time (these values would be `dx/dt_s, dy/dt_s, dz/dt_s)`) so we have to do the chain rule to recover the derivative in regular time. If `t_s = (t - baseTime) / timeScale` then `dt_s/dt = 1 / timeScale` and `dx/dt = dx/dt_s * dt_s/dt = dx/dt_s * 1 / timeScale`.

Reply:
@jessemapel Is this happening for both the position and velocity values?

Reply:
Just the velocity need to account for this.

Reply:
Talked about this more over Zoom but since the velocity isn't being time scaled at all the time itself doesn't need to account for scaling. No need to worry about removing it